File: lib/core/api_client.dart
import 'dart:async';
import 'dart:convert';
import 'package:dio/dio.dart';
import 'models.dart';

class AiApiClient {
  AiApiClient() : _dio = Dio(BaseOptions(connectTimeout: const Duration(seconds: 25), receiveTimeout: const Duration(minutes: 3)));
  final Dio _dio;
  CancelToken? _cancelToken;

  void cancel() => _cancelToken?.cancel('Остановлено');

  Map<String, String> _headers(AiProvider provider, String key) => switch (provider.protocol) {
    ApiProtocol.anthropic => {'x-api-key': key, 'anthropic-version': '2023-06-01', 'content-type': 'application/json'},
    ApiProtocol.gemini => {'content-type': 'application/json'},
    ApiProtocol.openAi => {'authorization': 'Bearer $key', 'content-type': 'application/json', if (provider.id == 'openrouter') 'HTTP-Referer': 'https://nitronbox.app', if (provider.id == 'openrouter') 'X-Title': 'NitronBox'},
  };

  String _custom(String base, String endpoint) {
    final clean = base.trim().replaceFirst(RegExp(r'/+$'), '');
    final uri = Uri.tryParse(clean);
    if (uri == null || !uri.hasScheme || !{'http', 'https'}.contains(uri.scheme)) throw Exception('Некорректный Base URL');
    if (clean.endsWith('/chat/completions')) return endpoint == 'models' ? clean.replaceFirst(RegExp(r'/chat/completions$'), '/models') : clean;
    return '$clean${clean.endsWith('/v1') ? '' : '/v1'}/$endpoint';
  }

  Future<List<AiModel>> loadModels(AiProvider provider, String key, String baseUrl) async {
    var url = provider.custom ? _custom(baseUrl, 'models') : provider.modelsUrl;
    if (provider.protocol == ApiProtocol.gemini) url = '$url?key=${Uri.encodeQueryComponent(key)}&pageSize=1000';
    try {
      final response = await _dio.get<Object?>(url, options: Options(headers: _headers(provider, key)));
      final body = response.data is Map ? Map<String, dynamic>.from(response.data! as Map) : jsonDecode(response.data.toString()) as Map<String, dynamic>;
      final source = (provider.protocol == ApiProtocol.gemini ? body['models'] : body['data'] ?? body['models']) as List? ?? const [];
      final result = source.whereType<Map>().where((raw) {
        if (provider.protocol != ApiProtocol.gemini) return true;
        return (raw['supportedGenerationMethods'] as List? ?? const []).contains('generateContent');
      }).map((raw) {
        final item = Map<String, dynamic>.from(raw);
        var id = (item['id'] ?? item['name'] ?? '').toString();
        if (provider.protocol == ApiProtocol.gemini) id = id.replaceFirst('models/', '');
        return AiModel(id, (item['displayName'] ?? item['name'] ?? id).toString(), (item['description'] ?? item['owned_by'] ?? '').toString());
      }).where((model) => model.id.isNotEmpty).toList()..sort((a, b) => a.name.toLowerCase().compareTo(b.name.toLowerCase()));
      return result;
    } on DioException catch (error) { throw Exception(_dioError(error)); }
  }

  Stream<String> streamChat({required AiProvider provider, required String key, required String baseUrl, required String model, required String system, required double temperature, required List<ChatMessage> messages}) async* {
    _cancelToken = CancelToken();
    var url = provider.custom ? _custom(baseUrl, 'chat/completions') : provider.chatUrl;
    if (provider.protocol == ApiProtocol.gemini) url = '$url/models/${Uri.encodeComponent(model)}:streamGenerateContent?alt=sse&key=${Uri.encodeQueryComponent(key)}';
    final data = switch (provider.protocol) {
      ApiProtocol.openAi => {'model': model, 'stream': true, 'temperature': temperature, 'messages': [if (system.trim().isNotEmpty) {'role': 'system', 'content': system.trim()}, ...messages.map((m) => {'role': m.role, 'content': m.content})]},
      ApiProtocol.anthropic => {'model': model, 'stream': true, 'max_tokens': 4096, 'temperature': temperature, if (system.trim().isNotEmpty) 'system': system.trim(), 'messages': messages.map((m) => {'role': m.role, 'content': m.content}).toList()},
      ApiProtocol.gemini => {if (system.trim().isNotEmpty) 'systemInstruction': {'parts': [{'text': system.trim()}]}, 'contents': messages.map((m) => {'role': m.role == 'assistant' ? 'model' : 'user', 'parts': [{'text': m.content}]}).toList(), 'generationConfig': {'temperature': temperature}},
    };
    try {
      final response = await _dio.post<ResponseBody>(url, data: data, cancelToken: _cancelToken, options: Options(headers: _headers(provider, key), responseType: ResponseType.stream));
      final lines = response.data!.stream
          .cast<List<int>>()
          .transform(utf8.decoder)
          .transform(const LineSplitter());
      await for (final line in lines) {
        if (!line.startsWith('data:')) continue;
        final payload = line.substring(5).trim();
        if (payload.isEmpty || payload == '[DONE]') continue;
        final event = jsonDecode(payload) as Map<String, dynamic>;
        if (event['error'] case final Map error) throw Exception(error['message'] ?? 'Ошибка потока');
        final delta = switch (provider.protocol) {
          ApiProtocol.openAi => (((event['choices'] as List?)?.firstOrNull as Map?)?['delta'] as Map?)?['content']?.toString() ?? '',
          ApiProtocol.anthropic => event['type'] == 'content_block_delta' ? ((event['delta'] as Map?)?['text']?.toString() ?? '') : '',
          ApiProtocol.gemini => _geminiText(event),
        };
        if (delta.isNotEmpty) yield delta;
      }
    } on DioException catch (error) {
      if (!CancelToken.isCancel(error)) throw Exception(_dioError(error));
    } finally { _cancelToken = null; }
  }

  String _geminiText(Map<String, dynamic> event) {
    final candidates = event['candidates'] as List?;
    if (candidates == null || candidates.isEmpty) return '';
    final content = (candidates.first as Map?)?['content'] as Map?;
    final parts = content?['parts'] as List? ?? const [];
    return parts.whereType<Map>().map((part) => part['text']?.toString() ?? '').join();
  }

  String _dioError(DioException error) {
    final data = error.response?.data;
    if (data is Map) return ((data['error'] as Map?)?['message'] ?? data['message'] ?? 'Ошибка ${error.response?.statusCode}').toString();
    return error.message ?? 'Не удалось подключиться к провайдеру';
  }
}

extension<T> on List<T> { T? get firstOrNull => isEmpty ? null : first; }

File: android/app/src/main/java/app/nitronbox/mobile/ApiClient.kt
package app.nitronbox.mobile

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.Call
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONArray
import org.json.JSONObject
import java.util.concurrent.TimeUnit

class ApiClient {
    private val jsonType = "application/json; charset=utf-8".toMediaType()
    private val client = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(3, TimeUnit.MINUTES)
        .build()
    @Volatile private var activeCall: Call? = null

    fun cancel() { activeCall?.cancel() }

    suspend fun models(provider: Provider, apiKey: String, baseUrl: String): List<AiModel> = withContext(Dispatchers.IO) {
        val endpoint = when {
            provider.custom -> modelsUrl(baseUrl)
            provider.protocol == Protocol.GEMINI -> "${provider.modelsUrl}?key=${apiKey.encode()}&pageSize=1000"
            else -> provider.modelsUrl
        }
        val request = requestBuilder(endpoint, provider, apiKey).get().build()
        val call = client.newCall(request)
        activeCall = call
        call.execute().use { response ->
            val raw = response.body?.string().orEmpty()
            val body = raw.jsonOrEmpty()
            if (!response.isSuccessful) throw ApiException(errorMessage(body, response.code))
            val source = if (provider.protocol == Protocol.GEMINI) body.optJSONArray("models") else body.optJSONArray("data") ?: body.optJSONArray("models")
            val result = buildList {
                for (index in 0 until (source?.length() ?: 0)) {
                    val item = source!!.optJSONObject(index) ?: continue
                    if (provider.protocol == Protocol.GEMINI) {
                        val methods = item.optJSONArray("supportedGenerationMethods")
                        if (methods != null && !(0 until methods.length()).any { methods.optString(it) == "generateContent" }) continue
                    }
                    val rawId = item.optString("id").ifBlank { item.optString("name") }
                    val id = if (provider.protocol == Protocol.GEMINI) rawId.removePrefix("models/") else rawId
                    if (id.isNotBlank()) add(AiModel(id, item.optString("displayName").ifBlank { item.optString("name").ifBlank { id } }, item.optString("description").ifBlank { item.optString("owned_by") }))
                }
            }
            result.distinctBy { it.id }.sortedBy { it.title.lowercase() }
        }.also { activeCall = null }
    }

    suspend fun chat(
        provider: Provider,
        apiKey: String,
        baseUrl: String,
        model: String,
        system: String,
        temperature: Float,
        messages: List<ChatMessage>,
        onDelta: (String) -> Unit,
    ) = withContext(Dispatchers.IO) {
        val endpoint = when {
            provider.custom -> chatUrl(baseUrl)
            provider.protocol == Protocol.GEMINI -> "${provider.chatUrl}/models/${model.encode()}:streamGenerateContent?alt=sse&key=${apiKey.encode()}"
            else -> provider.chatUrl
        }
        val payload = when (provider.protocol) {
            Protocol.ANTHROPIC -> anthropicPayload(model, system, temperature, messages)
            Protocol.GEMINI -> geminiPayload(system, temperature, messages)
            Protocol.OPENAI -> openAiPayload(model, system, temperature, messages)
        }
        val request = requestBuilder(endpoint, provider, apiKey)
            .post(payload.toString().toRequestBody(jsonType)).build()
        val call = client.newCall(request)
        activeCall = call
        call.execute().use { response ->
            if (!response.isSuccessful) {
                val body = response.body?.string().orEmpty().jsonOrEmpty()
                throw ApiException(errorMessage(body, response.code))
            }
            val source = response.body?.source() ?: throw ApiException("Провайдер вернул пустой ответ")
            while (!source.exhausted()) {
                val line = source.readUtf8Line() ?: break
                if (!line.startsWith("data:")) continue
                val data = line.removePrefix("data:").trim()
                if (data.isEmpty() || data == "[DONE]") continue
                val event = data.jsonOrEmpty()
                event.optJSONObject("error")?.let { throw ApiException(it.optString("message", "Ошибка потока")) }
                val delta = when (provider.protocol) {
                    Protocol.ANTHROPIC -> if (event.optString("type") == "content_block_delta") event.optJSONObject("delta")?.optString("text").orEmpty() else ""
                    Protocol.GEMINI -> event.optJSONArray("candidates")?.optJSONObject(0)?.optJSONObject("content")?.optJSONArray("parts")?.strings("text").orEmpty()
                    Protocol.OPENAI -> event.optJSONArray("choices")?.optJSONObject(0)?.optJSONObject("delta")?.optString("content").orEmpty()
                }
                if (delta.isNotEmpty()) onDelta(delta)
            }
        }
        activeCall = null
    }

    private fun requestBuilder(url: String, provider: Provider, key: String): Request.Builder {
        val builder = Request.Builder().url(url).header("Accept", "application/json")
        when (provider.protocol) {
            Protocol.ANTHROPIC -> builder.header("x-api-key", key).header("anthropic-version", "2023-06-01")
            Protocol.GEMINI -> Unit
            Protocol.OPENAI -> builder.header("Authorization", "Bearer $key")
        }
        if (provider.id == "openrouter") builder.header("HTTP-Referer", "https://nitronbox.app").header("X-Title", "NitronBox")
        return builder
    }

    private fun openAiPayload(model: String, system: String, temperature: Float, messages: List<ChatMessage>) = JSONObject().apply {
        put("model", model); put("stream", true); put("temperature", temperature)
        put("messages", JSONArray().apply {
            if (system.isNotBlank()) put(JSONObject().put("role", "system").put("content", system))
            messages.forEach { put(JSONObject().put("role", it.role).put("content", it.content)) }
        })
    }

    private fun anthropicPayload(model: String, system: String, temperature: Float, messages: List<ChatMessage>) = JSONObject().apply {
        put("model", model); put("stream", true); put("max_tokens", 4096); put("temperature", temperature)
        if (system.isNotBlank()) put("system", system)
        put("messages", JSONArray().apply { messages.forEach { put(JSONObject().put("role", it.role).put("content", it.content)) } })
    }

    private fun geminiPayload(system: String, temperature: Float, messages: List<ChatMessage>) = JSONObject().apply {
        if (system.isNotBlank()) put("systemInstruction", JSONObject().put("parts", JSONArray().put(JSONObject().put("text", system))))
        put("contents", JSONArray().apply { messages.forEach { put(JSONObject().put("role", if (it.role == "assistant") "model" else "user").put("parts", JSONArray().put(JSONObject().put("text", it.content)))) } })
        put("generationConfig", JSONObject().put("temperature", temperature))
    }

    private fun modelsUrl(base: String): String {
        val clean = base.trim().trimEnd('/')
        require(clean.startsWith("http://") || clean.startsWith("https://")) { "Некорректный Base URL" }
        return if (clean.endsWith("/chat/completions")) clean.removeSuffix("/chat/completions") + "/models" else clean + if (clean.endsWith("/v1")) "/models" else "/v1/models"
    }

    private fun chatUrl(base: String): String {
        val clean = base.trim().trimEnd('/')
        require(clean.startsWith("http://") || clean.startsWith("https://")) { "Некорректный Base URL" }
        return if (clean.endsWith("/chat/completions")) clean else clean + if (clean.endsWith("/v1")) "/chat/completions" else "/v1/chat/completions"
    }

    private fun String.encode() = java.net.URLEncoder.encode(this, "UTF-8")
    private fun String.jsonOrEmpty() = try { JSONObject(this) } catch (_: Exception) { JSONObject() }
    private fun errorMessage(body: JSONObject, code: Int) = body.optJSONObject("error")?.optString("message")?.takeIf { it.isNotBlank() } ?: body.optString("message").takeIf { it.isNotBlank() } ?: "Ошибка провайдера $code"
    private fun JSONArray.strings(key: String) = buildString { for (i in 0 until length()) append(optJSONObject(i)?.optString(key).orEmpty()) }
}

class ApiException(message: String) : Exception(message)
